src/files/BinaryFile.cpp

```diff
@@ -0,0 +1,745 @@
+        );
+    }
+#endif
+}
+
+
+uint64_t BinaryFile::tell() const {
+#if CHEMFILES_BINARY_FILE_USE_MMAP
+    return offset_;
+#else
+    return static_cast<uint64_t>(ftell64(file_));
+#endif
+}
+
+
+void BinaryFile::seek(uint64_t position) {
+#if CHEMFILES_BINARY_FILE_USE_MMAP
+    offset_ = position;
+#else
+    fseek64(file_, static_cast<off64_t>(position), SEEK_SET);
+#endif
+}
+
+
+/******************************************************************************/
+
+#define CHEMFILES_LITTLE_ENDIAN 0
+#define CHEMFILES_BIG_ENDIAN 1
+
+/***** adapted from boost ******/
+// GNU libc offers the helpful header <endian.h> which defines
+// __BYTE_ORDER
+#if defined (__GLIBC__)
+# include <endian.h>
+# if (__BYTE_ORDER == __LITTLE_ENDIAN)
+#  define CHEMFILES_BYTE_ORDER CHEMFILES_LITTLE_ENDIAN
+# elif (__BYTE_ORDER == __BIG_ENDIAN)
+#  define CHEMFILES_BYTE_ORDER CHEMFILES_BIG_ENDIAN
+# else
+#  error Unknown machine endianness detected.
+# endif
+#elif (defined(_BIG_ENDIAN) || defined(__BIG_ENDIAN__)) && !(defined(_LITTLE_ENDIAN) || defined(__LITTLE_ENDIAN__))
+# define CHEMFILES_BYTE_ORDER CHEMFILES_BIG_ENDIAN
+#elif (defined(_LITTLE_ENDIAN) || defined(__LITTLE_ENDIAN__)) && !(defined(_BIG_ENDIAN) || defined(__BIG_ENDIAN__))
+# define CHEMFILES_BYTE_ORDER CHEMFILES_LITTLE_ENDIAN
+#elif defined(__sparc) || defined(__sparc__) \
+   || defined(_POWER) || defined(__powerpc__) \
+   || defined(__ppc__) || defined(__hpux) || defined(__hppa) \
+   || defined(_MIPSEB) || defined(_POWER) \
+   || defined(__s390__)
+# define CHEMFILES_BYTE_ORDER CHEMFILES_BIG_ENDIAN
+#elif defined(__i386__) || defined(__alpha__) \
+   || defined(__ia64) || defined(__ia64__) \
+   || defined(_M_IX86) || defined(_M_IA64) \
+   || defined(_M_ALPHA) || defined(__amd64) \
+   || defined(__amd64__) || defined(_M_AMD64) \
+   || defined(__x86_64) || defined(__x86_64__) \
+   || defined(_M_X64) || defined(__bfin__)
+# define CHEMFILES_BYTE_ORDER CHEMFILES_LITTLE_ENDIAN
+#else
+# error Unknown target machine endianness
+#endif
+/***** adapted from boost ******/
+
+inline uint16_t swap_u16(uint16_t value) {
+    return static_cast<uint16_t>(
+        ((static_cast<uint32_t>(value) & 0xff) << 8) |
+        ((static_cast<uint32_t>(value) >> 8) & 0xff)
+    );
+}
+
+inline uint32_t swap_u32(uint32_t value) {
+    return (
+        (value << 24) |
+        ((value <<  8) & 0x00ff0000) |
+        ((value >>  8) & 0x0000ff00) |
+        ((value >> 24) & 0x000000ff)
+    );
+}
+
+inline uint64_t swap_u64(uint64_t value) {
+    return (
+        ((value & 0x00000000000000FFULL) << 56) |
+        ((value & 0x000000000000FF00ULL) << 40) |
+        ((value & 0x0000000000FF0000ULL) << 24) |
+        ((value & 0x00000000FF000000ULL) <<  8) |
+        ((value & 0x000000FF00000000ULL) >>  8) |
+        ((value & 0x0000FF0000000000ULL) >> 24) |
+        ((value & 0x00FF000000000000ULL) >> 40) |
+        ((value & 0xFF00000000000000ULL) >> 56)
+    );
+}
```

Comment:
It just came to my mind that there might be builtin functions for swapping. And it turns out [there are](https://stackoverflow.com/a/46137633) (e.g. [glibc](https://sourceware.org/git/?p=glibc.git;a=blob;f=bits/byteswap.h;h=ca8417d59a80d3db8ac00a866b76cb3cd98a97f0;hb=HEAD)). This is also utilized by [libxtc](https://gitlab.com/impulse_md/libxtc/-/blob/58f40e1e9d392ec607e775772658eb1b748be549/libxtc/xtc_unpack.cpp#L54-96).
I'm not sure if it's worth bothering though because from my [preliminary fooling around](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEEgCcpAAOqAqETgwe3r7%2ByanpAiFhkSwxcVyJdpgOGUIETMQEWT5%2BAVU1AnUNBEUR0bEJtvWNzTltwz2hfaUDFQCUtqhexMjsHOYAzKHI3lgA1CYbbsgKBPiCh9gmGgCC1zdeoQQAbJIA%2BgR7CgDuTElvXleEEegleHz2ADcxF5MHMDgB2Ky3PYovbETAEZYMPYQe6o/E4iBQ7yYA5mZ57DSqDQ02l0ukAMQZciCQThhzcHL2AFZnuz4ZzkQSUVBiTCyRSqfTpRomTSWWyDkcuZINPzBTdhSKidDSeZJdSZTS5bSFezlUc9mZJOq8VrRbqJZTDdKTXSzUrOZa9gAOW1C4UOklOqX0t30j2XS4ov0IjVawlivXk52MhlG%2BWs81XDbYK02uN2wM64P61Oy9MZzOKqO5vaq/2a%2B0l8VlqnhquR3PR3lzIt7PsbJF3eEAEXuE7MWwYOy8%2Bw5UQAngRMD8/gA6BBR24gl7vT7AWh4KLIN5RNf/IG7sGfYn84f49GY4jY8%2B/S%2BSEuD4cmMccBa0Jw3K8H4HBaKQqCcJyljWF8SwrMmGw8KQBCaP%2BCwANYgNyGj6JwkggWhEGcLwCggLhqFgf%2BpBwLASBoCwSR0LE5CUAxTH0HExAQq8AC0wDIMgXAaGYuE0LQK7EGREBRERUShA0i6cMhDFsIIADyDC0EpVGkFgLCGIeazgfg6I1BCq5EZgqjVF4K7KbwTyYIBulHlExCKR4WBEQQxB4CwDkLFQBjAAoABqeCYN86lJIwDkyIIIhiOwUgJfIShqERuhcPohkoNY1j6MeZGQAsqBJI4AhkRwvHqRspHOdUlV%2BBArijH4OXBFMJRlHoKRpM17V9fkzW9D1sy2I1HQMF0IyeC0ejtM1s2TMU/TlEM3RDTlpzdGN60SAsCjwasei%2BZgaw8ABQGEbpkEcAASgAkkIbi8WFaI8ZIAl7AJyB7MJ65mOuGg4tBViWKQey4IQJBkkhUMeIxzHEPDXBzLwlFaHMmHYbhLkEaQoHgfdpHkShaE43hHBmLdJMkRTVFUxZUkZCAkhAA%3D%3D), it seems there is only a difference in the assembler for MIPS and RISC architectures.

Reply:
> It just came to my mind that there might be builtin functions for swapping. And it turns out [there are](https://stackoverflow.com/a/46137633) (e.g. [glibc](https://sourceware.org/git/?p=glibc.git;a=blob;f=bits/byteswap.h;h=ca8417d59a80d3db8ac00a866b76cb3cd98a97f0;hb=HEAD)).

I find it easier to just have the same implementation everywhere instead of using `#ifdef` everywhere to find the right one =)

> I'm not sure if it's worth bothering though because from my [preliminary fooling around](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEEgCcpAAOqAqETgwe3r7%2ByanpAiFhkSwxcVyJdpgOGUIETMQEWT5%2BAVU1AnUNBEUR0bEJtvWNzTltwz2hfaUDFQCUtqhexMjsHOYAzKHI3lgA1CYbbsgKBPiCh9gmGgCC1zdeoQQAbJIA%2BgR7CgDuTElvXleEEegleHz2ADcxF5MHMDgB2Ky3PYovbETAEZYMPYQe6o/E4iBQ7yYA5mZ57DSqDQ02l0ukAMQZciCQThhzcHL2AFZnuz4ZzkQSUVBiTCyRSqfTpRomTSWWyDkcuZINPzBTdhSKidDSeZJdSZTS5bSFezlUc9mZJOq8VrRbqJZTDdKTXSzUrOZa9gAOW1C4UOklOqX0t30j2XS4ov0IjVawlivXk52MhlG%2BWs81XDbYK02uN2wM64P61Oy9MZzOKqO5vaq/2a%2B0l8VlqnhquR3PR3lzIt7PsbJF3eEAEXuE7MWwYOy8%2Bw5UQAngRMD8/gA6BBR24gl7vT7AWh4KLIN5RNf/IG7sGfYn84f49GY4jY8%2B/S%2BSEuD4cmMccBa0Jw3K8H4HBaKQqCcJyljWF8SwrMmGw8KQBCaP%2BCwANYgNyGj6JwkggWhEGcLwCggLhqFgf%2BpBwLASBoCwSR0LE5CUAxTH0HExAQq8AC0wDIMgXAaGYuE0LQK7EGREBRERUShA0i6cMhDFsIIADyDC0EpVGkFgLCGIeazgfg6I1BCq5EZgqjVF4K7KbwTyYIBulHlExCKR4WBEQQxB4CwDkLFQBjAAoABqeCYN86lJIwDkyIIIhiOwUgJfIShqERuhcPohkoNY1j6MeZGQAsqBJI4AhkRwvHqRspHOdUlV%2BBArijH4OXBFMJRlHoKRpM17V9fkzW9D1sy2I1HQMF0IyeC0ejtM1s2TMU/TlEM3RDTlpzdGN60SAsCjwasei%2BZgaw8ABQGEbpkEcAASgAkkIbi8WFaI8ZIAl7AJyB7MJ65mOuGg4tBViWKQey4IQJBkkhUMeIxzHEPDXBzLwlFaHMmHYbhLkEaQoHgfdpHkShaE43hHBmLdJMkRTVFUxZUkZCAkhAA%3D%3D), it seems there is only a difference in the assembler for MIPS and RISC architectures.

Both MIPS and RISC contains a direct call to the intrinsic function (`call    __bswapdi2`) instead of inlining it inside `glibc_bswap_64`, and this function should contain the same assembly as `swap_u64`.

Reply:
Having the same implementation everywhere makes sense:+1: